font/src/directwrite/mod.rs

```diff
@@ -13,33 +13,207 @@
+use std::borrow::Cow;
+use std::collections::HashMap;
+
 extern crate dwrote;
 use self::dwrote::{
     FontCollection, FontStretch, FontStyle, FontWeight, GlyphOffset, GlyphRunAnalysis,
 };
 
+extern crate winapi;
+use self::winapi::um::dwrite;
+
 use super::{
-    BitmapBuffer, FontDesc, FontKey, GlyphKey, Metrics, RasterizedGlyph, Size, Slant, Style, Weight,
+    BitmapBuffer, FontDesc, FontKey, GlyphKey, Metrics, Rasterize, RasterizedGlyph, Size, Slant,
+    Style, Weight,
 };
 
+mod locale;
+use self::locale::get_current_locale;
+
+pub struct TextAnalysisSource {
+    locale: String,
+    length: u32,
+}
+
+impl dwrote::TextAnalysisSourceMethods for TextAnalysisSource {
+    fn get_locale_name(&self, _text_position: u32) -> (Cow<str>, u32) {
+        (Cow::Borrowed(&self.locale), self.length)
+    }
+
+    fn get_paragraph_reading_direction(&self) -> dwrite::DWRITE_READING_DIRECTION {
+        dwrite::DWRITE_READING_DIRECTION_LEFT_TO_RIGHT
+    }
+}
+
+pub struct LoadedFontFace {
+    font: dwrote::Font,
+    font_face: dwrote::FontFace,
+}
+
 pub struct DirectWriteRasterizer {
-    fonts: Vec<dwrote::FontFace>,
+    fonts: HashMap<FontKey, LoadedFontFace>,
+    keys: HashMap<FontDesc, FontKey>,
     device_pixel_ratio: f32,
+    font_collection: dwrote::FontCollection,
 }
 
-impl crate::Rasterize for DirectWriteRasterizer {
-    type Err = Error;
+impl DirectWriteRasterizer {
+    fn rasterize_glyph_with_fallback(&mut self, glyph: GlyphKey) -> Result<RasterizedGlyph, Error> {
+        self.rasterize_glyph(glyph).or_else(|error| {
+            if let Error::MissingGlyph(_) = &error {
+                // Try to get a fallback font
+                let font = &self.fonts.get(&glyph.font_key).unwrap().font;
+                let locale = get_current_locale();
+                let utf16_codepoints = encode_utf16(glyph.c);
+                let num_codepoints = utf16_codepoints.len() as u32;
+
+                let fallback = dwrote::FontFallback::get_system_fallback().unwrap();
+                let fallback_font = fallback.map_characters(
+                    &dwrote::TextAnalysisSource::from_text_and_number_subst(
+                        Box::new(TextAnalysisSource {
+                            locale: locale.clone(),
+                            length: num_codepoints,
+                        }),
+                        utf16_codepoints,
+                        dwrote::NumberSubstitution::new(
+                            dwrite::DWRITE_NUMBER_SUBSTITUTION_METHOD_NONE,
+                            &locale,
+                            false,
+                        ),
+                    ),
+                    0,
+                    num_codepoints,
+                    &self.font_collection,
+                    Some(&font.family_name()),
+                    font.weight(),
+                    font.style(),
+                    font.stretch(),
+                );
+
+                if let Some(mapped_font) = fallback_font.mapped_font {
```

Comment:
What if `mapped_font` is `None`? When would this be the case? Currently we explode right now, so we might as well make that explicit by returning the error directly from this point, rather than making it look like we're performing error correction?